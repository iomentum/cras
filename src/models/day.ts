/*

         Day

         / \
     worked holy

*/
export class Day {
  date: Date;

  constructor(date:Date){
    this.date = date;
  }

  totalWorked():number {return 0;}
  getDayDate():Date {return this.date}

}

export class WorkedDay extends Day {
  morning: boolean;
  afternoon: boolean;

  constructor(date: Date) {
    super(date);
    this.morning = false;
    this.afternoon = false;
  }

  addMorning(): void {
    this.morning = true;
  }

  removeMorning(): void {
    this.morning = false;
  }

  addAfternoon(): void {
    this.afternoon = true;
  }

  removeAfternoon(): void {
    this.afternoon = false;
  }

  addWholeDay(): void {
    this.addMorning();
    this.addAfternoon();
  }

  reset(): void {
    this.removeMorning();
    this.removeAfternoon();
  }

  totalWorked(): number {
    return (this.morning ? 0.5 : 0) + (this.afternoon ? 0.5 : 0);
  }
}
// un jour peut etre travaillable ou pas.
// si il est travaillable, une personne peut travailler: toute la journee, une demi journee, ou pas.

export class Holiday extends Day {}