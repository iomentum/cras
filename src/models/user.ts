export class User {
  firstName: string;
  lastName:string;
  customer: string;
  email: string;
  uid: string;
  signatureURL: string;

  constructor (firstName:string,lastName:string,customer: string,email: string,uid: string) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.customer = customer;
    this.email = email;
    this.uid = uid;
    this.signatureURL = '';
  }

  setInfos (firstName:string, lastName:string, customer: string, email:string, uid:string): void {
    this.firstName = firstName;
    this.lastName = lastName;
    this.customer = customer;
    this.email = email;
    this.uid = uid;
  }

  setSignatureURL(url:string): void {
    this.signatureURL = url;
  }

  setFirstName(username:string): void {
    this.firstName = username;
  }

  setCustomer(customer:string): void {
    this.customer = customer;
  }

  resetSignatureURL(): void {
    this.signatureURL = "";
  }
}