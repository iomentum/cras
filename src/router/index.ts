import { createWebHistory, createRouter } from "vue-router";
import MainView from '@/views/MainView.vue';
import PrintView from '@/views/PrintView.vue';

const routes = [
  {
    path: "/",
    name: "Home",
    component: MainView
  },
  {
    path: "/print",
    name: "PrintView",
    component: PrintView
  }
];

const router = createRouter({
  history: createWebHistory(),
  routes
});

export default router;
