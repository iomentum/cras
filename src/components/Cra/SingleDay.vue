<script setup lang="ts">
import { Day, WorkedDay, Holiday } from '@/models/day';
import { useDaysStore } from '@/stores/daysStore';
import { useRoute } from 'vue-router';
import { computed } from '@vue/reactivity';

const route = useRoute();
const store = useDaysStore();

// eslint-disable-next-line no-undef
const props = defineProps<{ day: WorkedDay | Day | Holiday }>();

const toggleHalfDay = (evt: Event) => {
  let target = evt.target as HTMLInputElement;
  if(props.day instanceof WorkedDay) {
    store.toggleHalfDay(props.day, target.checked, target.name);
  }
}

const isMainView = computed(() => {
  return route.path == "/"
})

</script>

<template>
  <div
    v-if="isMainView"
    class="day"
  >
    <div>{{ props.day.getDayDate().getDate() }}</div>
    <div
      v-if="(day instanceof WorkedDay)"
      class="checkboxes"
    >
      <input
        type="checkbox"
        name="morning"
        :checked="day.morning"
        @change="toggleHalfDay"
      >
      <input
        type="checkbox"
        name="afternoon"
        :checked="day.afternoon"
        @change="toggleHalfDay"
      >
    </div>
    <div
      v-else
      class="holiday"
    />
  </div>
  <div
    v-else
    class="day"
  >
    <div>{{ props.day.getDayDate().getDate() }}</div>
    <div
      v-if="(day instanceof WorkedDay)"
      class="total-worked-day"
    >
      <p>{{ props.day.totalWorked() }}</p>
    </div>
    <div
      v-else
      class="holiday"
    />
  </div>
</template>

<style scoped lang="scss">
.day {
  display: flex;
  flex-direction: column;
  width: 45px;
  border: 1px solid $main-color;
  border-right: 0;
  color: $main-color;

  & p {
    font-family: "bau-regular", Arial, Helvetica, sans-serif;
  }

  & .total-worked-day {
    height: 59%;
  }

  & .checkboxes {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    margin-left: 15px;
    justify-content: space-around;
  }

  & div:nth-child(1) {
    height: 40%;
    border-bottom: 1px solid $main-color;
  }

  & .holiday {
    background-color: #edeaea;
    height: 59%;
  }
}
</style>