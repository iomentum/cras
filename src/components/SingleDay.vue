<script setup lang="ts">
import { Day } from '@/models/day'
import { useDaysStore } from '@/stores/daysStore'
import { isWeekend } from '@/utils/utils';
import { useRoute } from 'vue-router'

// eslint-disable-next-line no-undef
const props = defineProps<{ day: Day }>()
const route = useRoute()
const store = useDaysStore();
const thisDay = props.day.date.getDate()-1

function isMainView(){
  if(route.name == 'Home'){
    return true
  } else {
    return false
  }
}

function toggle(event: Event){
  const target = event.target as HTMLInputElement
  store.changeDayProps(thisDay, target.name)
}

function changeOvertime(event: Event){
  const target = event.target as HTMLInputElement
  store.changeOvertime(thisDay, parseInt(target.value))
}

function counter(prop:string){
  let count = 0
  switch(prop){
    case "workedDay":{
      if (props.day.workedDay.morning) count+=0.5
      if (props.day.workedDay.afternoon) count+=0.5
      return count
    }
    case "vacationDay":{
      if (props.day.vacationDay.morning) count+=0.5
      if (props.day.vacationDay.afternoon) count+=0.5
      return count
    }
    case "holiday":{
      if (props.day.holiday) count = 1
      else count = 0
      return count
    }
  }
}
</script>

<template>
  <div :class="[{ weekend: isWeekend(day) },{ mainview: isMainView()},'day']">
    <div>
      {{ day.date.getDate() }}
    </div>
    <div v-if="!isWeekend(day) && isMainView()">
      <input
        type="checkbox"
        name="morning"
        :checked="day.workedDay.morning"
        @change="toggle"
      >
      <input
        type="checkbox"
        name="afternoon"
        :checked="day.workedDay.afternoon"
        @change="toggle"
      >
    </div>
    <div v-else>
      <input
        v-if="!isWeekend(day) && counter('workedDay')!=0"
        type="text"
        :value="counter('workedDay')"
        disabled="true"
      >
    </div>
    <div v-if="!isWeekend(day) && isMainView()">
      <input
        type="checkbox"
        name="holiday"
        :checked="day.holiday"
        @change="toggle"
      >
    </div>
    <div v-else>
      <input
        v-if="counter('holiday') != 0"
        type="text"
        name="holiday"
        :value="counter('holiday')"
        disabled="true"
      >
    </div>
    <div v-if="!isWeekend(day) && isMainView()">
      <input
        type="checkbox"
        name="vacationDayMorning"
        :checked="day.vacationDay.morning"
        @change="toggle"
      >
      <input
        type="checkbox"
        name="vacationDayAfternoon"
        :checked="day.vacationDay.afternoon"
        @change="toggle"
      >
    </div>
    <div v-else>
      <input
        v-if="counter('vacationDay') != 0"
        type="text"
        :value="counter('vacationDay')"
        disabled="true"
      >
    </div>
    <div
      v-if="!isWeekend(day) && isMainView()"
      class="input-overtime"
    >
      <input
        type="number"
        :value="day.overTime"
        @change="changeOvertime"
      >
    </div>
    <div
      v-else
      class="input-overtime"
    >
      <input
        v-if="day.overTime != 0"
        :value="day.overTime"
        disabled="true"
      >
    </div>
  </div>
</template>

<style scoped lang="scss">
.weekend{
  background-color: #EDEAEA;
}
.mainview div{
  &:nth-child(2n+1){
    height: 16%;
  }
  &:nth-child(2n){
    height: 26%;
  }
}
.day {
  display: flex;
  flex-direction: column;
  width: 38px;
  text-align: center;
  font-family: "bau-regular";
  & > div {
    padding: 5px;
    border: 1px solid $main-color;
    box-sizing: border-box;
    height: 20%;
    & input[type=text]{
      border: none;
      width: 100%;
      -moz-appearance: textfield;
      text-align: center;
      background-color : #ffffff00;
    }
  }
  & :first-child {
    font-weight: bold;
    background-color: white;
  }
  & .input-overtime {
    text-align: center;
    & input {
      border: none;
      width: 100%;
      -moz-appearance: textfield;
      text-align: center;
      background-color : #ffffff00;
    }
  }
}
.input input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
  padding: 0;
}
.input[type=text] input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
  padding: 0;
}
</style>