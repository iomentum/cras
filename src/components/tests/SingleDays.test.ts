import { mount } from '@vue/test-utils'
import { WorkedDay, Holiday, Day } from '@/models/day'
import SingleDay from '@/components/SingleDay.vue'
import { setActivePinia, createPinia } from 'pinia';
import { createTestingPinia } from '@pinia/testing'
import { createApp } from 'vue';
import router from '@/router/index';

const workedDay = new WorkedDay(new Date(2022,4,11))

describe('TableDays unit test', () => {
  const app = createApp({})
  beforeEach(() => {
    const pinia = createPinia()
    app.use(pinia)
    setActivePinia(pinia)
  });

  const wrapper = mount(SingleDay,{
    props: {
      day: workedDay
    },
    global: {
      plugins: [createTestingPinia(), router],
    }
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
});
