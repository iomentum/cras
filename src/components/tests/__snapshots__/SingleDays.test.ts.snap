// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableDays unit test renders correctly 1`] = `
<div
  class="day"
>
  <div>
    10
  </div>
  <div
    class="checkboxes"
  >
    <input
      name="morning"
      type="checkbox"
    />
    <input
      name="afternoon"
      type="checkbox"
    />
  </div>
</div>
`;
